### DISCUSSION TITLE
Sometimes getting `Missing or invalid client id.`

### DISCUSSION THREAD

**Discussion by MarvinJWendt** [pocketbase/pocketbase#1150] [Q&A] [answered]
Hi, we use the realtime API, but sometimes (every ~2 out of 5 times) we get this error: `Missing or invalid client id.`

<img width="558" alt="image" src="https://user-images.githubusercontent.com/31022056/204846000-9d3163be-a789-4e2f-97ba-9beddfa27c99.png">

If the error comes up, we have to reload the page a couple of times to make it work again.

What could be the reason for that?

And is there any way to catch this error, or maybe auto-reconnect if it happens?

```js
    client.collection("products").subscribe('*', (data) => {
        console.log("Products changed:", data)
        // ...
    }).then((unsub) => {
        console.log("Subscribed to products")
    }).catch((err) => {
        console.error("Error subscribing to products", err)
    })
```

The `catch` is never called, the error directly "panics" to the console.

**Comment by ganigeorgiev** [comment, marked as answer]
This error is usually thrown on `POST /api/realtime` and it could be triggered if:

- the subscription call is made with a client id different than the one that is returned with the initial PB_CONNECT event (you could inspect the payload of `POST /api/realtime`)

- the client is sending a "submit subscriptions" call but the server crashed/restarted before the request was able to reach the backend (we keep the client id only in memory, aka. as part of the app instance)

SSE in the browser has autoreconnect. In normal situations you should be able to `catch` the error if it happened during the `subscribe()` call, but I think in your case the error is thrown inside the `EventSource` listener either on init or reconnect, because internally we do something like:

```js
eventSource.addEventListener('PB_CONNECT', (e) => {
    ... do something else...

    submitSubscriptions();
});
```

I'll consider in future versions to wrap it and rethrow it to a global realtime `onError` callback or similar (and maybe pass a promise from the `subscribe()` call to the `PB_CONNECT` listener and resolve/reject depending on the `EventSource` init/error state).

---

But I'm not entirely sure what is causing it to fail in your case. You can try debugging it by starting the PocketBase executable with `--debug` and watch the "Realtime connection" logs in the stdout.

**Reply by ganigeorgiev** [reply]
@MarvinJWendt 

**A fix to allow catching the realtime connect error(s) was added in [JS SDK v0.8.2](https://github.com/pocketbase/js-sdk/releases/tag/v0.8.2).**

It ended a little more complicated than I initially anticipated (_I'm planning a rewrite of the realtime SDK handling in the near future and I will replace the browser's `EventSource` with a streamed `fetch` which should simplify a little the code and make it a little more easy to test_). How it currently works:

When you first try to establish an SSE connection, we queue all `subscribe()` calls and will resolve the subscribe promise(s) only after successful SSE connect AND subscriptions submission.
Or in other words, in case of a connect error, all multiple subscribe catch callbacks will be triggered but internally we still will maintain a single SSE connection:

```js
// all errors are wrapped in ClientResponseError
// you can get the actual error by inspecting err.originalError
pb.collection('users').subscribe('*').catch((err) => console.log('sub 1 error handling'));
pb.collection('users').subscribe('*').catch((err) => console.log('sub 2 error handling'));
```

If a connection error occur **after** the SSE connection was established, a custom stepped autoretry mechanism will be initialized (_it is custom because the native `EventSource` auto reconnect handling turned out not very reliable and it is not consistent across the browsers and the node polyfill implementations_).

The `catch` is not triggered on auto retry because otherwise it could block the execution indefinitely, especially if `await` was used. Additionally, usually the autoretry will happen long after the initial successful `subscribe()` call and it doesn't make much sense to interrupt/block the flow in this case since the reconnect executes seamlessly in the background.

There are other various caveats to the current implementation due to the numerous async edge cases. For example calling `subscribe()` during **pending reconnect** could trigger eventually an additional `submitSubscriptions()` call in case the new subscribe adds new topics that differs from the already sent ones:

```
interrupted---->reconnectAttempt---->connected--->submitSubscriptions--X-->success
                                                                       ^
                                                                       | someone called another subscribe()
                                                                       | check if there are new topics and resubmit if necessary
```
